sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/query/ExpressionParser.java

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package io.stargate.sgv2.docsapi.service.query;
+
+import com.bpodgursky.jbool_expressions.And;
+import com.bpodgursky.jbool_expressions.Expression;
+import com.bpodgursky.jbool_expressions.Literal;
+import com.bpodgursky.jbool_expressions.Not;
+import com.bpodgursky.jbool_expressions.Or;
+import com.bpodgursky.jbool_expressions.rules.RuleSet;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.Streams;
+import io.stargate.sgv2.docsapi.api.common.properties.document.DocumentProperties;
+import io.stargate.sgv2.docsapi.api.exception.ErrorCode;
+import io.stargate.sgv2.docsapi.api.exception.ErrorCodeRuntimeException;
+import io.stargate.sgv2.docsapi.service.query.ImmutableFilterExpression.Builder;
+import io.stargate.sgv2.docsapi.service.query.condition.BaseCondition;
+import io.stargate.sgv2.docsapi.service.query.condition.ConditionParser;
+import io.stargate.sgv2.docsapi.service.query.filter.operation.FilterHintCode;
+import io.stargate.sgv2.docsapi.service.util.DocsApiUtils;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+
+public class ExpressionParser {
```

Comment:
Should be annotated with `@ApplicationScoped` or `@Singleton` so it can be injected in other components and/or tests.